### PR TITLE
Add documentation for http_hsts.md

### PR DESCRIPTION
**Pull request by tekwizz123**
Add in documentation for the auxiliary/scanner/http/http_hsts module. Should fix one of the modules listed at #12389.

### CODE REVIEW

**Comment by h00die** [comment]
Please run `tools/dev/msftidy_docs.rb`.  It's a new tool to standardize documentation, it showed a few errors which should be easy to correct.

**Comment by tekwizz123** [comment]
@h00die Ah okay thats odd, cause I modeled this off of the http_header.md documentation, and it seems that doesn't conform to the standards used by `tools/dev/msftidy_docs.rb`. Similarly I can see a lot of the other scanner modules also don't conform to the standards used by `tools/dev/msftidy_docs.rb`.

Anyway just wanted to point that out but agree with making the documentation standard so will go ahead and change so long. Seems the main thing I was missing was the ## Options section and then placing the ## Vulnerable Application section first in the documentation followed by ### Description.

Will upload changes shortly.

**Comment by tekwizz123** [comment]
@h00die Added commits. Not 100% sure if the ##Options section is correct, as this module doesn't really have any options specific to it, so tried to model it off auxiliary/scanner/http/git_scanner.md. Hope that helps!

**Comment by h00die** [comment]
`msftidy_docs.rb` is a relatively recent invention, and there have been some changes to how docs are structured.  Currently there are 4,723 issues with docs in the framework.  While I've put in some effort to try to get to the backdated ones (#13133, #12878, #12831) it is a struggle.  Ones moving forward should conform to the new standard.  If you'd like to help correct some old docs to conform, I won't argue!!!

If there are no module specific options, feel free to just leave an empty section.  @tperry-r7 did we agree to just have a h2 and h3s if there are no module specific options?  I think we did, but just confirming.

In the future `msftidy_docs.rb` may start notifying to remove 'standard' options.  @tperry-r7 agree to that? slack me and we can start a list.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+## Verification Steps
+
+1. Do: ```use auxiliary/scanner/http/http_hsts```
+2. Do: ```set RHOSTS [IP]```
+3. Do: ```set RPORT [PORT]```
+4. Do: ```run```
+
+## Options
+
+  **Proxies**
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/http/http_hsts.md]: as per comments, you can remove all the options, just leave the h2 header.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
+## Vulnerable Application
+
+### Description
+
+This module checks to see whether or not the scanned systems return the HSTS header to enforce HSTS.
+
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/http/http_hsts.md]: did you set the server up yourself? if so, can you provide instructions on how to do so on apache/nginx (just one is fine).

File: documentation/modules/auxiliary/scanner/http/http_hsts.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  **THREADS**
+  Number of threads to use to complete the scan. Default is `1`.
+
+  **VHOST**
+  Virtual host of the HTTP servers being scanned. Default is ``.
+
+## Scenarios
+
+**Running the scanner**
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/http/http_hsts.md]: This should be an H3 that says what you are scanning (ie ### Apache 2.2 on Ubuntu 18.04)

File: documentation/modules/auxiliary/scanner/http/http_hsts.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+msf5 auxiliary(scanner/http/http_hsts) > set RHOSTS 192.168.24.20
+RHOSTS => 192.168.24.20
+msf5 auxiliary(scanner/http/http_hsts) > run
+
+[+] 192.168.24.20:443 - Strict-Transport-Security:max-age=31536000; includeSubDomains
+[*] Scanned 1 of 1 hosts (100% complete)
+[*] Auxiliary module execution completed
+msf5 auxiliary(scanner/http/http_hsts) >
+
+```
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/http/http_hsts.md]: If there is a tool that can verify this other than metasploit, I like to include it as well.

Something like https://github.com/rapid7/metasploit-framework/pull/12450/files#diff-b2c1aa63724b140259932dfb390a0711R356

https://nmap.org/nsedoc/scripts/http-security-headers.html

Only, I think it should be an H3 (@tperry-r7 agree?)

**Review by h00die — CHANGES_REQUESTED** [review]

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 1 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: Fixed this now with latest push.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 2 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: Sorry didn't set the server up myself, but I'll look into spinning up an Apache instance with this enabled.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 3 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: Sorry to be clear should **Running the scanner** should be changed to a H3 with the target that was scanned?

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 4 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: Should I try run another tool and include the output as a H3 header here? I know Nmap should have a NSE script that will provide similar information so can include that output if needed.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 3 above)
Comment (h00die) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: You can remove that, it should look more like https://github.com/rapid7/metasploit-framework/pull/12423/files#diff-d2659524190bc3d103650ebcd869e3e0R101

**Comment by tekwizz123** [comment]
> 
> 
> `msftidy_docs.rb` is a relatively recent invention, and there have been some changes to how docs are structured. Currently there are 4,723 issues with docs in the framework. While I've put in some effort to try to get to the backdated ones (#13133, #12878, #12831) it is a struggle. Ones moving forward should conform to the new standard. If you'd like to help correct some old docs to conform, I won't argue!!!
> 
> If there are no module specific options, feel free to just leave an empty section. @tperry-r7 did we agree to just have a h2 and h3s if there are no module specific options? I think we did, but just confirming.
> 
> In the future `msftidy_docs.rb` may start notifying to remove 'standard' options. @tperry-r7 agree to that? slack me and we can start a list.

Happy to help with backfixing old cases as well! Though sounds like given your comments here there may need to be some updates to the `msftidy_docs.rb` script r.e removing standard options so not sure if that should be done first?

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 4 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: Sorry ignore my last comment, I see what you mean now :)

Comment (h00die) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: I usually just have 1 other tool if its free and on Kali (aka common).  It helps for when youre on a test and one tool says one thing, a 2nd tool says something else, and you need a tie breaker.  Not a requirement, but honestly `nmap -p 80 -sV --script <abc> <ip>` won't take but 3min to add.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 1 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: @h00die Can confirm if this is good now? If so will mark this as resolved.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 3 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: Pushing an update which should fix this in a sec.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 2 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: Got the Apache server set up, next commit should add instructions on how to do so.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 4 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md]: Next commit should have the output from running the http-security-headers script in Nmap on the Apache host that I set up.

**Comment by tekwizz123** [comment]
@h00die Made changes, should now be ready for another round of review; tried to replicate formatting but it may be a bit off in some places so let me know if you think any areas should be altered.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md [hunk 5]
```diff
@@ -0,0 +1,296 @@
+## Vulnerable Application
+
+### Description
+
+This module checks to see whether or not the scanned systems return the HSTS header to enforce HSTS.
+
+### Install on Ubuntu
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/http/http_hsts.md:7]: which ubuntu, 18.04?

File: documentation/modules/auxiliary/scanner/http/http_hsts.md [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+1. Do: ```use auxiliary/scanner/http/http_hsts```
+2. Do: ```set RHOSTS [IP]```
+3. Do: ```set RPORT [PORT]```
+4. Do: ```run```
+
+## Options
+
+## Scenarios
+
+### Apache 2.4.29 on Ubuntu 18.03 LTS
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/http/http_hsts.md:244]: 18.03 or 18.04?

File: documentation/modules/auxiliary/scanner/http/http_hsts.md [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+Install using following instructions for Ubuntu listed above.
+
+```
+msf5 > use auxiliary/scanner/http/http_hsts
+msf5 auxiliary(scanner/http/http_hsts) > show options
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/http/http_hsts.md:250]: you can drop the show options and output since you didn't change any

**Review by h00die — CHANGES_REQUESTED** [review]
A few super minor things, and this is ready.  The biggest thing though is why is `apache_mod_cgi_bash_env.md` now included in this PR?  I don't remember seeing that before, but that file already exists in framework.

**Comment by tekwizz123** [comment]
> 
> 
> A few super minor things, and this is ready. The biggest thing though is why is `apache_mod_cgi_bash_env.md` now included in this PR? I don't remember seeing that before, but that file already exists in framework.

Ah sorry new to doing updates to this, not sure why that is included, I didn't do any edits to that file. Is there any way I can get git to drop that file/discard the edits?

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 6 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md:244]: Ah that should be 18.04, my bad will resolve that.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 5 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md:7]: Yeah was meant to be 18.04, will fix that up so long.

File: documentation/modules/auxiliary/scanner/http/http_hsts.md (hunk 7 above)
Comment (tekwizz123) [reply documentation/modules/auxiliary/scanner/http/http_hsts.md:250]: Sounds good, will do :)

**Comment by h00die** [comment]
looks good to me, I'll land this shortly

**Comment by h00die** [comment]
I had to do some toying to get `msftidy_docs` and the actual view in the browser all happy.  https://github.com/rapid7/metasploit-framework/commit/02e909c6b26f5fdcc7a6ec63563aa70d002c7871

**Comment by h00die** [comment]
Thanks for the docs!!! Thanks for sticking with it and adding all those additional things like sever setup, and nmap.

**Comment by h00die** [comment]
# Release Notes
This PR adds docs for the http_hsts aux scanner module.

**Comment by tekwizz123** [comment]
@h00die No problem :) For future what changes did you have to make? I know msftidy was asking me to indent the code, which it seems you did with the last change, but I'm a bit confused as to why the &lt; changes were needed?

**Comment by h00die** [comment]
All the changes I did were here: https://github.com/rapid7/metasploit-framework/commit/02e909c6b26f5fdcc7a6ec63563aa70d002c7871

When doing the `info -d` in metasploit, when the ``` was there with indents, it was including them in the block.  The < was also being interpreted as html (even though I told it not to), so I did the `&lt;` to avoid the processing.  Unfortunately it was just a bunch of try this, try that, try another thing, to get it to look correct.

The big thing is just to run `info -d` to catch anything getting wonky in the md to html conversion.

**Comment by tekwizz123** [comment]
> 
> 
> All the changes I did were here: [02e909c](https://github.com/rapid7/metasploit-framework/commit/02e909c6b26f5fdcc7a6ec63563aa70d002c7871)
> 
> When doing the `info -d` in metasploit, when the ``` was there with indents, it was including them in the block. The < was also being interpreted as html (even though I told it not to), so I did the `&lt;` to avoid the processing. Unfortunately it was just a bunch of try this, try that, try another thing, to get it to look correct.
> 
> The big thing is just to run `info -d` to catch anything getting wonky in the md to html conversion.

Thanks @h00die, didn't know about the `info -d` command, that makes a lot of sense! Will keep that in mind for when I do future documentation contributions :)

**Pull request merged**